src/MongoDB.Driver.Core/Core/Authentication/ScramSha1Authenticator.cs

```diff
@@ -36,294 +37,43 @@ public sealed class ScramSha1Authenticator : SaslAuthenticator
-                using (var sha1 = SHA1.Create())
-                {
-                    return sha1.ComputeHash(data);
-                }
-            }
-
-            private static byte[] Hi(string password, byte[] salt, int iterations)
-            {
-                return new Rfc2898DeriveBytes(
-                    password,
-                    salt,
-                    iterations).GetBytes(20); // this is length of output of a sha-1 hmac
-            }
+            var passwordDigest = AuthenticationHelper.MongoPasswordDigest(credential.Username, credential.Password);
+            // 20 is the length of output of a sha-1 hmac
+            return new Rfc2898DeriveBytes(passwordDigest, salt, iterations).GetBytes(20);
+        }
 
-            private static byte[] HMAC(UTF8Encoding encoding, byte[] data, string key)
-            {
+        private static byte[] Hmac1(UTF8Encoding encoding, byte[] data, string key)
+        {
 #if NET45
-                using (var hmac = new HMACSHA1(data, useManagedSha1: true))
+            using (var hmac = new HMACSHA1(data, useManagedSha1: true))
 #else
                 using (var hmac = new HMACSHA1(data))
 #endif
-                {
-                    return hmac.ComputeHash(encoding.GetBytes(key));
-                }
-            }
-        }
-
-        private class ClientLast : ISaslStep
-        {
-            private readonly byte[] _bytesToSendToServer;
-            private readonly byte[] _serverSignature64;
-
-            public ClientLast(byte[] bytesToSendToServer, byte[] serverSignature64)
             {
-                _bytesToSendToServer = bytesToSendToServer;
-                _serverSignature64 = serverSignature64;
-            }
-
-            public byte[] BytesToSendToServer
-            {
-                get { return _bytesToSendToServer; }
-            }
-
-            public bool IsComplete
-            {
-                get { return false; }
-            }
-
-            public ISaslStep Transition(SaslConversation conversation, byte[] bytesReceivedFromServer)
-            {
-                var encoding = Utf8Encodings.Strict;
-                var map = NVParser.Parse(encoding.GetString(bytesReceivedFromServer));
-                var serverSignature = Convert.FromBase64String(map['v']);
-
-                if (!ConstantTimeEquals(_serverSignature64, serverSignature))
-                {
-                    throw new MongoAuthenticationException(conversation.ConnectionId, message: "Server signature was invalid.");
-                }
-
-                return new CompletedStep();
-            }
-
-            private bool ConstantTimeEquals(byte[] a, byte[] b)
-            {
-                var diff = a.Length ^ b.Length;
-                for (var i = 0; i < a.Length && i < b.Length; i++)
-                {
-                    diff |= a[i] ^ b[i];
-                }
-
-                return diff == 0;
-            }
-        }
-
-        private class NVParser
```

Comment:
Again, moved to a new class so that the functionality could be shared between SCRAM-SHA-1/256 Authenticators.